### PR TITLE
Separate toolkit modules

### LINKED ISSUE #834: Split Monolithic Files
**Is your feature request related to a problem? Please describe.**

There are several files in the project which are over 5000 lines long, including `toolkits.py` and `molecule.py`. Having such large files make a) IDEs which perform static code analysis run slowly when editing the files, especially in terms of code completion and linting, b) makes it hard to find what you're looking for without scrolling over 1000s of lines, c) makes it daunting for potential new contributors.

It would be good to split these up into smaller sub-files. The module imports can be retained by exposing classes in the split files via the root `__init__.py` of a module so as not to break the API.

**Describe the solution you'd like**

As a start, split the following would be a big plus:

```
- utils
    - toolkits
        - __init__.py         <- Expose all classes through here to not break API.
            | GLOBAL_TOOLKIT_REGISTRY
            | OPENEYE_AVAILABLE         <- Consider removing these as I'm not sure what they actually do?
            | RDKIT_AVAILABLE
            | AMBERTOOLS_AVAILABLE
            | BASIC_CHEMINFORMATICS_TOOLKITS  <- Remove the warning as it's never triggered?
        - exceptions.py         <- Maybe add a module for the general custom exceptions?
            | MissingPackageError         <- Delete as it is unused?
            | ToolkitUnavailableException
            | LicenseError
            | InvalidToolkitError
            | InvalidToolkitRegistryError
            | UndefinedStereochemistryError
            | ChargeMethodUnavailableError
            | IncorrectNumConformersError
            | IncorrectNumConformersWarning
            | ChargeCalculationError
            | InvalidIUPACNameError
        - toolkits.py
            | ToolkitWrapper
            | BuiltInToolkitWrapper
            | ToolkitRegistry
        - ambertools.py
            | GAFFAtomTypeWarning
            | AntechamberNotFoundError
        - rdkit.py
            | RDKitToolkitWrapper
        - openeye.py
            | OpenEyeToolkitWrapper
```

similar attention should likely be given to `molecule.py` and potentially `topology.py`.

### PR DESCRIPTION
**Pull request by adalke**
This PR breaks up the monolithic "toolkits.py" into several submodules:
```
ambertools_wrapper.py
base_wrapper.py
builtin_wrapper.py
constants.py
exceptions.py
openeye_wrapper.py
rdkit_wrapper.py
toolkit_registry.py
```

- [x] Tag issue being addressed

 "Split Monolithic Files" https://github.com/openforcefield/openff-toolkit/issues/834 . See also related discussion in "move each toolkit wrapper implementation into its own module"   https://github.com/openforcefield/openff-toolkit/issues/988

- [x] Resolves #834

- [ ] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/master/openff/toolkit/tests)

No tests added - this is a straight re-factoring.

- [x] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/master/docs), if applicable

A couple of docstrings were chaned to use r"aw" strings because they contained unsupported \escapes and generated warnings during test.

- [x] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase

Black and isort both pass. I added some relative imports, so relabelled "=== GLOBAL IMPORTS ===" to "=== IMPORTS ==="

- [ ] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/master/docs/releasehistory.rst)

Is this needed?

### CODE REVIEW

**Comment by adalke** [comment]
I moved all of the exceptions into its own module. Some of the existing code refers to openff.topology.toolkit.util.utils like:
```
./toolkit/topology/molecule.py:
from openff.toolkit.utils.utils import MissingDependencyError, requires_package

./toolkit/tests/test_utils.py:
from openff.toolkit.utils.utils import MissingDependencyError, requires_package

./toolkit/typing/engines/smirnoff/parameters.py: 
from openff.toolkit.utils.utils import (
    IncompatibleUnitError,
    MessageException,
     ...
```

I have not changed those places. I assume they should be using `openff.toolkit.utils` instead of `openff.toolkit.utils.utils` and that the current location is not meant as part of the public API so can be changed.

**Comment by mattwthompson** [comment]
@j-wags would this be an appropriate time to reconsider the module structure? Personal preference, but I'm not a huge fan of core functionality tucked away into `something.utils.something`. Something like `from openff.toolkit.wrappers import RDKitToolkitWrapper` feels better to me, and presumably things could be wired up to point in the right places in order to maintain backwards API compatibility. (This is probably most suitable for a follow-up PR to this one). Totally fine if this would be considered too much of a change.

**Comment by adalke** [comment]
I agree with @mattwthompson both on the layout change, and the proposal that it be in another PR.

File: openff/toolkit/utils/toolkits.py
Comment (j-wags) [inline openff/toolkit/utils/toolkits.py:180]: (blocking) I'm not sure if this was accidentally committed or not.  Right now it makes it so that, if a developer added a new Exception type to `exceptions.py` and updated that file's `__all__`, they'd get an error unless they also added the new Exception to _this_ file's `__all__`. Is that the intent? 

If it's unintentional, I'm ok with this check being deleted

If it's intentional:

* Is it worth having this check at all / do we WANT to force this file's `__all__` to be a superset of all the imported files `__all__`s? 
* Would be be better (or possible) to put this in `test_toolkits` instead of here?
* If this is the pattern we want, would this be generally satisfiable by using `from exceptions import *`?

**Review by j-wags — CHANGES_REQUESTED** [review]
I think this looks pretty good @adalke! There's an added code block in here that I thought about for a while, but either I don't understand it or it's unintentional, so let's resolve that before merging. 

One neat formatting feature -- In the description text for this PR, you can add a line like `- [x] Resolves #834`, and github will automatically close the issue when this PR gets merged, leaving a link to the PR itself for future records.

> A couple of docstrings were changed to use r"aw" strings because they contained unsupported \escapes and generated warnings during test.

Excellent. Thanks! 

> >     Update changelog
> Is this needed?

You're right -- It's not.

> I assume they should be using openff.toolkit.utils instead of openff.toolkit.utils.utils and that the current location is not meant as part of the public API so can be changed.

Hm, I haven't really thought about whether the longer import paths shouldn't be considered part of the public API. Stylistically I agree that importing from `X.utils.utils` in our code looks bad, and that we should use the shorter paths throughout the codebase. But generally, is there a situation where we'd benefit by considering the longer paths to be outside the public API?

> would this be an appropriate time to reconsider the module structure?

The big motivator to do that in this PR would be if that would change a lot of lines again. I'd like to minimize the huge-number-of-changed-lines PRs, since those require a lot of my time to meaningfully review. But as I understand it, switching the modules to a different folder would be file _moves_, not file _edits_, so the line diff shouldn't be huge (just some changed imports). So those can be a different PR

Also, the changes in that move may be overridden by a refactor from `ToolkitWrappers` to `Toolbox`es depending on how we choose to proceed with @adalke's proposals. So let's hold off on that until we have the other moving pieces settled.

**Comment by adalke** [comment]
> There's an added code block in here that I thought about for a while, but either I don't understand it or it's unintentional, so let's resolve that before merging.

It's a mistake. It was there to ensure my __all__ list was correct. I left it there because of habits from being primarily a sole developer - I know that it's potentially useful, but potentially not useful, and if in the future it proves to be not useful then I can remove it.

However, in a group projects like OpenFF, that context isn't visible, and the way I did it looks like it's trying to enforce a long-term policy - which wasn't my intent. So I removed it.

 
>  you can add a line like `- [x] Resolves #834`, and github will automatically close the issue when this PR gets merged, leaving a link to the PR itself for future records.

Updated.

> But generally, is there a situation where we'd benefit by considering the longer paths to be outside the public API?

Refactors like this.

The internal layout of a package should be the one that's easiest to maintain. This might not be the same as the logical layout. Unfortunately, Python makes it hard to create a module in place X but make it's repr(), etc. follow the logical layout.

If you want to refactor the internals in the future, you'll break code which reaches into the deeper structure. For example, reaching into toolbox.utils.utils for an exception, when now there's a toolbox.exceptions.

**Review by j-wags — APPROVED** [review]
> It's a mistake. It was there to ensure my all list was correct.

Awesome. Yeah, @mattwthompson and @jchodera have caught a lot of my unintentional commits and I'm really grateful for their time - The system works!

>>    But generally, is there a situation where we'd benefit by considering the longer paths to be outside the public API?
> Refactors like this.

Oh, of course! Thanks for clarifying that. A lot of how we've been discussing the API just "clicked" in my head. 

This should be good to merge!

**Pull request merged**
